xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+                    }
+                    // todo: deal with invalide surrotage sequences
+                }
+            }
+        }
+        else if ( HEAD_CHECK(0xF8, 0xF0) ) {
+            if ( HAS_FOLLOWUP(3) && IS_FOLLOWING(1) && IS_FOLLOWING(2) && IS_FOLLOWING(3) ) {
+                if ( do_decode )
+                    *p = HEAD_BYTE(0x07, 18) | CONT_BYTE(1,12) | CONT_BYTE(2,6) | CONT_BYTE(3,0);
+                s += 4;
+                valid = true;
+            }
+        }
+        // else: invalid first byte in UTF-8 sequence
+
+        if ( !valid ) {
+            if (do_decode)
+                *p = REPLACEMENT_CHAR;
+            s++;
+            is_valid = false;
+        }
+        if ( do_decode ) {
+	    // Try to detect if we have RTL chars, so that if we don't have any,
+	    // we don't need to invoke expensive fribidi processing.
+	    if ( !has_rtl )
+                has_rtl = is_unicodepoint_rtl(*p);
+            p++;
+        }
+        ucount++;
+    }
+    return ucount;
+}
+
+// ==============================================
+// Flags, data structures, and global variables
+
+#define HINT_DIRECTION_IS_RTL   0x0001 /// segment direction is RTL
+#define HINT_BEGINS_PARAGRAPH   0x0002 /// segment is at start of paragraph
+#define HINT_ENDS_PARAGRAPH     0x0004 /// segment is at end of paragraph
+
+#define CHAR_CAN_WRAP_AFTER        0x0001
+#define CHAR_MUST_BREAK_AFTER      0x0002
+#define CHAR_SKIP_ON_BREAK         0x0004
+#define CHAR_CAN_EXTEND_WIDTH      0x0008
+#define CHAR_IS_CLUSTER_TAIL       0x0010
+#define CHAR_IS_RTL                0x0020
+#define CHAR_SCRIPT_CHANGE         0x0040
+#define CHAR_IS_PARA_START         0x0100
+#define CHAR_IS_PARA_END           0x0200
+#define CHAR_PARA_IS_RTL           0x0400 /// to know the line with this char is part
+                                           /// of a paragraph with main dir RTL
+
+// Info, after measure(), about each m_text char
+typedef struct {
+    unsigned short flags;
+    signed short   width;
+} xtext_charinfo_t;
+
+// Glyph info when shaping a line (to be returned to Lua as a table of tables)
+// (16 bytes, making our static s_shape_result[MAX_LINE_GLYPHS] a 128Kb buffer)
+typedef struct {
+    int text_index;    // original index in m_text
+    uint16_t glyph;    // glyph index in font
+    unsigned char font_num;
+    unsigned is_rtl:1;
+    unsigned can_extend:1;
+    unsigned _unused:6;
+    signed short  x_advance;
+    signed short  x_offset;
+    signed short  y_offset;
+    unsigned char cluster_len;
+    unsigned is_cluster_start:1;
+} xtext_shapeinfo_t;
+
+// Holder of HB data structures per font, to be stored as a userdata
+// in the Lua font table
+typedef struct {
+    hb_font_t *    hb_font;
+    hb_buffer_t *  hb_buffer;
+    hb_feature_t * hb_features;
+    int            hb_features_nb;
+} xtext_hb_font_data;
+
+// Global direction and language
+static bool default_para_direction_rtl = false;
+static char * default_lang = NULL;
```

Comment:
Random comment before I forget: Don't know where/how you use that (and other plain `char`s yet), but I assume you'd want `unsigned char`s pretty much everywhere in this context, to avoid surprises on platforms where char is signed.

Reply:
I thought that was when I was using it to store numbers (so issue, with signed and unsigned when I stored a negative number). There, I just store some string ("fr", "zh_CN"..)
Harfbuzz and FriBidi all have just `const char *` in their function declaration. Using something different, I would need to cast it, or something.

Reply:
Yup, made sense once I saw how it was used ;).